File: cpp/soluc.cpp
#include <chrono>
#include <iostream>
#include <fstream>

using namespace std;
using namespace std::chrono;

#define SIZE 5000  // Large matrix size

int main() {
    // Allocate a 2D array dynamically (avoid stack overflow)
    int** matrix = new int*[SIZE];
    for (int i = 0; i < SIZE; i++)
        matrix[i] = new int[SIZE];

    int num = 1;

    // Start time measurement
    auto start = high_resolution_clock::now();

    // Fill the matrix diagonally
    for (int col = 0; col < SIZE; col++) {
        int i = 0, j = col;
        while (i < SIZE && j >= 0) {
            matrix[i][j] = num++;
            i++;
            j--;
        }
    }

    for (int row = 1; row < SIZE; row++) {
        int i = row, j = SIZE - 1;
        while (i < SIZE && j >= 0) {
            matrix[i][j] = num++;
            i++;
            j--;
        }
    }

    // End time measurement
    auto stop = high_resolution_clock::now();

    // Compute elapsed time in microseconds
    auto duration = duration_cast<microseconds>(stop - start);

    // Save execution time to file
    ofstream file("/app/output/time.txt");
    if (file.is_open()) {
        file << duration.count() << " µs" << endl;
        file.close();
    }

    // Free allocated memory
    for (int i = 0; i < SIZE; i++)
        delete[] matrix[i];
    delete[] matrix;

    return 0;
}

File: nodejs/soluc.js
const fs = require('fs');

const SIZE = 5000; // Large matrix size

// Create an empty matrix (2D array)
const matrix = Array.from({ length: SIZE }, () => Array(SIZE).fill(0));

/**
 * Recursive function to fill the matrix diagonally
 * @param {number[][]} matrix - The 2D array to be filled
 * @param {number} i - Current row index
 * @param {number} j - Current column index
 * @param {object} num - Number object to keep track of the counter (passed by reference)
 */
function fillDiagonal(matrix, i, j, num) {
    if (i >= SIZE || j < 0) return; // Base case: Stop when out of bounds
    matrix[i][j] = num.value++;
    fillDiagonal(matrix, i + 1, j - 1, num); // Recursive step
}

// Start time measurement
const startTime = process.hrtime.bigint();

let num = { value: 1 }; // Using an object to pass by reference

// Fill the upper triangle of the matrix
for (let col = 0; col < SIZE; col++) {
    fillDiagonal(matrix, 0, col, num);
}

// Fill the lower triangle of the matrix
for (let row = 1; row < SIZE; row++) {
    fillDiagonal(matrix, row, SIZE - 1, num);
}

// End time measurement
const endTime = process.hrtime.bigint();
const timeTaken = Number((endTime - startTime) / 1000n); // Convert nanoseconds to microseconds

// Save execution time to file
fs.writeFileSync('/app/output/time.txt', `nodejs: ${timeTaken} µs\n`, 'utf8');
